### PR TITLE
core: Allow for multiple gnutls CA files (closes #569)

### LINKED ISSUE #569: Allow specifying multiple root CA files for GnuTLS
Right now to trust a root CA of some IRC server with custom certificate there are two options:
1. Append that certificate into system wide CA certs file (might not be accessible)
2. Create a new certs file which is a concatenation of the system's and your own (not easily kept in sync with the system's one)

This is why there should be a way to simply specify multiple files which contain CA certs to trust.

### PR DESCRIPTION
**Pull request by fceschmidt**
This has been tested with two CA files, the merged one from my default SSL certificate store and the Hackint certificate, which is also my use case for this feature.

For specifying multiple CA files in weechat.conf, separate CA files with a colon in between. They will be added to the trusted certificates at start-up time (upon the call to `network_set_gnutls_ca_file`). I additionally amended `string_expand_home` in such a fashion that it is now able to replace `~/` with `/home/...` not only at the beginning of the string, but also after each colon in the string.

I added some unit tests as well for the additional functionality of `string_expand_home`.

P.S.: Sorry for creating two pull requests, I should really embrace TDD...

### CODE REVIEW

**Comment by stfnm** [comment]
For splitting the strings at the colon you might also wanna consider using weechat's [string_split()](https://github.com/weechat/weechat/blob/master/src/core/wee-string.c#L1783), which could simplify the code a little further.

**Comment by fceschmidt** [comment]
Thanks for the hint, I did that and also fixed one more thing about Windows compatibility by adding a platform-dependent macro definition in `weechat.h`.

**Comment by sim642** [comment]
In my opinion `string_expand_home` function should not be dealing with separators in the case of multiple delimited paths. Wouldn't it simply be possible to achieve the same result with a lot less effort by calling `string_expand_home` on every one of the split parts?

File: src/core/wee-network.c [hunk 1]
```diff
@@ -107,9 +107,21 @@ network_set_gnutls_ca_file ()
...
+            char **single_paths;
+
+            single_paths = string_split(ca_path2, PATH_SEPARATOR, 0, 0, 
+                                        &num_paths);
+
+            for (i = 0; i < num_paths; i++)
+            {
+                gnutls_certificate_set_x509_trust_file (gnutls_xcred,
+                            single_paths[i], GNUTLS_X509_FMT_PEM);
+                free (single_paths[i]);
```
Comment (sim642) [inline src/core/wee-network.c]: Unnecessary, use `string_free_split` instead.

File: src/core/wee-network.c [hunk 2]
```diff
@@ -107,9 +107,21 @@ network_set_gnutls_ca_file ()
...
+                single_paths[i] = NULL;
+            }
             free (ca_path2);
+            free (single_paths);
```
Comment (sim642) [inline src/core/wee-network.c]: Use `string_free_split` here.

File: src/core/wee-network.c [hunk 3]
```diff
@@ -107,9 +107,21 @@ network_set_gnutls_ca_file ()
         ca_path2 = string_replace (ca_path, "%h", weechat_home);
         if (ca_path2)
         {
-            gnutls_certificate_set_x509_trust_file (gnutls_xcred, ca_path2,
-                                                    GNUTLS_X509_FMT_PEM);
+            int num_paths, i;
```
Comment (sim642) [inline src/core/wee-network.c]: It is a convention throughout WeeChat source to declare variables at the very beginning of a function.

**Comment by fceschmidt** [comment]
Just adding that this also provides a solution for #658

File: src/core/wee-network.c [hunk 4]
```diff
@@ -96,23 +96,30 @@ void
...
+    char *ca_path, *ca_path2, **single_paths;
+    int i;
 
     if (weechat_no_gnutls)
         return;
 
-    ca_path = string_expand_home (CONFIG_STRING(config_network_gnutls_ca_file));
-    if (ca_path)
+    single_paths = string_split(CONFIG_STRING(config_network_gnutls_ca_file), 
+                                PATH_SEPARATOR, 0, 0, NULL);
```
Comment (nimaje) [inline src/core/wee-network.c]: I don't think it is a good idea to have the seperator OS dependent, if it doesn't have to be like here

**Comment by flashcode** [comment]
This option has been replaced by `weechat.network.gnutls_ca_user` in WeeChat ≥ 3.2 and it supports several paths separated by `:`.

So I close this PR.

**Pull request closed without merging**